sparkmagic/sparkmagic/livyclientlib/livysession.py

```diff
@@ -14,14 +13,55 @@
     LivyUnexpectedStatusException, BadUserDataException
 
 
+class _HeartbeatThread(threading.Thread):
+    def __init__(self, livy_session, refresh_seconds, retry_seconds, run_at_most=None):
+        super(_HeartbeatThread, self).__init__()
+        
+        self.livy_session = livy_session
+        self.refresh_seconds = refresh_seconds
+        self.retry_seconds = retry_seconds
+        self.run_at_most = run_at_most
+
+    def run(self):
+        i = 0
+        if self.livy_session is not None:
+            self.livy_session.logger.info(u'Starting heartbeat for session {}'.format(self.livy_session.id))
+        else:
+            self.livy_session.logger.info(u'Will not start heartbeat because session is none')
+        
+        while self.livy_session is not None:
+            try:
+                self.livy_session.refresh_status()
+                sleep(self.refresh_seconds)
+            except Exception as e:
+                self.livy_session.logger.error(u'{}'.format(e))
+                sleep(self.retry_seconds)
+            
+            if self.run_at_most is not None:
+                i += 1
+                
+                if i >= self.run_at_most:
+                    return
+
+    def stop(self):
+        if self.livy_session is not None:
+            self.livy_session.logger.info(u'Stopping heartbeat for session {}'.format(self.livy_session.id))
+        
+        self.livy_session = None
+        self.join()
+
+
 class LivySession(ObjectWithGuid):
     def __init__(self, http_client, properties, ipython_display,
-                 session_id=-1, sql_created=None, spark_events=None):
+                 session_id=-1, sql_created=None, spark_events=None,
+                 should_heartbeat=False, heartbeat_thread=None):
         super(LivySession, self).__init__()
         assert u"kind" in list(properties.keys())
         kind = properties[u"kind"]
         self.properties = properties
         self.ipython_display = ipython_display
+        self._should_heartbeat = should_heartbeat
+        self._user_passed_heartbeat_thread = heartbeat_thread
```

Comment:
Why 2 different properties for the heartbeat thread? (`_user_passed_heartbeat_thread`, `_heartbeat_thread`)